File: android/core/core-content/src/main/kotlin/com/fortunepocket/core/content/BaziReadingGenerator.kt
package com.fortunepocket.core.content

import com.fortunepocket.core.model.BaziReading
import com.fortunepocket.core.model.ElementPersonality
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BaziReadingGenerator @Inject constructor(
    private val contentLoader: ContentLoader
) {

    fun generate(
        birthDateMillis: Long,
        birthHour: Int?
    ): BaziReading {
        val stems = contentLoader.loadHeavenlyStems().stems.sortedBy { it.index }
        val branches = contentLoader.loadEarthlyBranches().branches
        val templates = contentLoader.loadBaziTemplates()
        val isZh = Locale.getDefault().language == "zh"

        val birthDate = Instant.ofEpochMilli(birthDateMillis)
            .atZone(ZoneId.systemDefault())
            .toLocalDate()
        val referenceDate = LocalDate.of(1900, 1, 1)
        val daysBetween = ChronoUnit.DAYS.between(referenceDate, birthDate).toInt()
        val stem = stems[positiveModulo(daysBetween, stems.size)]
        val branch = branches.firstOrNull { it.month == birthDate.monthValue }
            ?: branches[positiveModulo(birthDate.monthValue - 1, branches.size)]

        val introPool = if (isZh) templates.overallIntroZh else templates.overallIntroEn
        val intro = if (introPool.isEmpty()) {
            ""
        } else {
            introPool[positiveModulo(daysBetween + branch.index, introPool.size)]
        }

        val phasePool = if (isZh) templates.phaseAdviceZh else templates.phaseAdviceEn
        val phaseSeed = (LocalDate.now().dayOfYear * 31) + (stem.index * 7) + (birthHour ?: 12)
        val elementSummary = elementDescription(
            templates.elementPersonality,
            stem.element,
            isZh
        )
        val monthNuance = if (isZh) branch.nuanceZh else branch.nuanceEn
        val combinedElementDescription = if (isZh) {
            "$elementSummary\n\n月令补充：$monthNuance"
        } else {
            "$elementSummary\n\nSeasonal nuance: $monthNuance"
        }

        return BaziReading(
            birthDate = birthDateMillis,
            birthHour = birthHour,
            stemId = stem.id,
            stemNameZh = stem.nameZh,
            stemNameEn = stem.nameEn,
            stemColorPrimary = stem.colorPrimary,
            disclaimer = if (isZh) templates.disclaimer_zh else templates.disclaimer_en,
            overallPersonality = listOf(
                intro,
                if (isZh) stem.personalityZh else stem.personalityEn
            ).filter { it.isNotBlank() }.joinToString("\n\n"),
            love = if (isZh) stem.loveZh else stem.loveEn,
            career = if (isZh) stem.careerZh else stem.careerEn,
            wealth = if (isZh) stem.wealthZh else stem.wealthEn,
            phaseAdvice = if (phasePool.isEmpty()) {
                ""
            } else {
                phasePool[positiveModulo(phaseSeed, phasePool.size)]
            },
            elementDescription = combinedElementDescription
        )
    }

    private fun elementDescription(
        elementPersonality: ElementPersonality,
        element: String,
        isZh: Boolean
    ): String {
        return when (element) {
            "wood" -> if (isZh) elementPersonality.wood_zh else elementPersonality.wood_en
            "fire" -> if (isZh) elementPersonality.fire_zh else elementPersonality.fire_en
            "earth" -> if (isZh) elementPersonality.earth_zh else elementPersonality.earth_en
            "metal" -> if (isZh) elementPersonality.metal_zh else elementPersonality.metal_en
            "water" -> if (isZh) elementPersonality.water_zh else elementPersonality.water_en
            else -> if (isZh) elementPersonality.wood_zh else elementPersonality.wood_en
        }
    }

    private fun positiveModulo(value: Int, mod: Int): Int {
        return ((value % mod) + mod) % mod
    }
}

File: ios/PocketOracle/Core/Services/BaziReadingGenerator.swift
import Foundation

enum BaziReadingGenerator {

    static func generate(
        birthDate: Date,
        birthHour: Int?
    ) throws -> BaziReading {
        let loader = ContentLoader.shared
        let stems = try loader.loadHeavenlyStems().stems.sorted { $0.index < $1.index }
        let branches = try loader.loadEarthlyBranches().branches
        let templates = try loader.loadBaziTemplates()
        let calendar = Calendar.current
        let isZh = AppLanguageOption.isChinese

        let startOfDay = calendar.startOfDay(for: birthDate)
        let referenceDate = calendar.date(from: DateComponents(year: 1900, month: 1, day: 1)) ?? startOfDay
        let daysBetween = calendar.dateComponents([.day], from: referenceDate, to: startOfDay).day ?? 0
        let stem = stems[positiveModulo(daysBetween, stems.count)]

        let birthMonth = calendar.component(.month, from: birthDate)
        let branch = branches.first(where: { $0.month == birthMonth })
            ?? branches[positiveModulo(birthMonth - 1, branches.count)]

        let introPool = isZh ? templates.overallIntroZh : templates.overallIntroEn
        let phasePool = isZh ? templates.phaseAdviceZh : templates.phaseAdviceEn
        let introSeed = positiveModulo(daysBetween + branch.index, max(introPool.count, 1))
        let phaseSeed = (Date().dayOfYear * 31) + (stem.index * 7) + (birthHour ?? 12)
        let intro = introPool.isEmpty ? "" : introPool[introSeed]
        let disclaimer = isZh ? templates.disclaimerZh : templates.disclaimerEn
        let monthNuance = isZh ? branch.nuanceZh : branch.nuanceEn
        let elementSummary = templates.elementPersonality.description(
            element: stem.element,
            isZh: isZh
        )
        let combinedElementDescription = isZh
            ? "\(elementSummary)\n\n月令补充：\(monthNuance)"
            : "\(elementSummary)\n\nSeasonal nuance: \(monthNuance)"

        return BaziReading(
            birthDate: birthDate,
            birthHour: birthHour,
            stem: stem,
            disclaimer: disclaimer,
            overallPersonality: [intro, stem.localizedPersonality]
                .filter { !$0.isEmpty }
                .joined(separator: "\n\n"),
            love: stem.localizedLove,
            career: stem.localizedCareer,
            wealth: stem.localizedWealth,
            phaseAdvice: phasePool.isEmpty
                ? ""
                : phasePool[positiveModulo(phaseSeed, phasePool.count)],
            elementDescription: combinedElementDescription
        )
    }

    private static func positiveModulo(_ value: Int, _ mod: Int) -> Int {
        ((value % mod) + mod) % mod
    }
}
